diff --git a/src/http/ngx_http_request.h b/src/http/ngx_http_request.h
index 1a4eae1..359fb5e 100644
--- a/src/http/ngx_http_request.h
+++ b/src/http/ngx_http_request.h
@@ -119,6 +119,7 @@
 #define NGX_HTTP_GATEWAY_TIME_OUT          504
 #define NGX_HTTP_INSUFFICIENT_STORAGE      507
 
+#define NGX_HTTP_QUEUE_EXPIRATION          555
 
 #define NGX_HTTP_LOWLEVEL_BUFFERED         0xf0
 #define NGX_HTTP_WRITE_BUFFERED            0x10
diff --git a/src/http/ngx_http_upstream.c b/src/http/ngx_http_upstream.c
index 3f77c03..eafb606 100644
--- a/src/http/ngx_http_upstream.c
+++ b/src/http/ngx_http_upstream.c
@@ -14,8 +14,6 @@ static void ngx_http_upstream_rd_check_broken_connection(ngx_http_request_t *r);
 static void ngx_http_upstream_wr_check_broken_connection(ngx_http_request_t *r);
 static void ngx_http_upstream_check_broken_connection(ngx_http_request_t *r,
     ngx_event_t *ev);
-static void ngx_http_upstream_connect(ngx_http_request_t *r,
-    ngx_http_upstream_t *u);
 static ngx_int_t ngx_http_upstream_reinit(ngx_http_request_t *r,
     ngx_http_upstream_t *u);
 static void ngx_http_upstream_send_request(ngx_http_request_t *r,
@@ -435,12 +433,12 @@ ngx_http_upstream_init(ngx_http_request_t *r)
 
 found:
 
-    if (uscf->peer.init(r, uscf) != NGX_OK) {
-        ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
-        return;
+    switch(uscf->peer.init(r, uscf)) {
+      case NGX_OK:   ngx_http_upstream_connect(r, u);
+      case NGX_BUSY: return;
     }
-
-    ngx_http_upstream_connect(r, u);
+    
+    ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
 }
 
 
@@ -534,9 +532,12 @@ ngx_http_upstream_check_broken_connection(ngx_http_request_t *r,
         return;
     }
 
+    /* max_connections patch */
+    /*
     if (u->peer.connection == NULL) {
         return;
     }
+    */
 
 #if (NGX_HAVE_KQUEUE)
 
@@ -639,7 +640,7 @@ ngx_http_upstream_check_broken_connection(ngx_http_request_t *r,
 }
 
 
-static void
+void
 ngx_http_upstream_connect(ngx_http_request_t *r, ngx_http_upstream_t *u)
 {
     ngx_int_t          rc;
diff --git a/src/http/ngx_http_upstream.h b/src/http/ngx_http_upstream.h
index 2ed2797..c73d136 100644
--- a/src/http/ngx_http_upstream.h
+++ b/src/http/ngx_http_upstream.h
@@ -270,6 +270,7 @@ ngx_int_t ngx_http_upstream_header_variable(ngx_http_request_t *r,
     ngx_http_variable_value_t *v, uintptr_t data);
 
 void ngx_http_upstream_init(ngx_http_request_t *r);
+void ngx_http_upstream_connect(ngx_http_request_t *r, ngx_http_upstream_t *u);
 ngx_http_upstream_srv_conf_t *ngx_http_upstream_add(ngx_conf_t *cf,
     ngx_url_t *u, ngx_uint_t flags);
 ngx_int_t ngx_http_upstream_hide_headers_hash(ngx_conf_t *cf,
