diff -Naur src/http/ngx_http_upstream.c ../nginx-0.6.32/src/http/ngx_http_upstream.c
--- src/http/ngx_http_upstream.c	2008-11-09 20:19:46.000000000 +0100
+++ ../nginx-0.6.32/src/http/ngx_http_upstream.c	2008-04-17 16:32:11.000000000 +0200
@@ -14,6 +14,8 @@
 static void ngx_http_upstream_wr_check_broken_connection(ngx_http_request_t *r);
 static void ngx_http_upstream_check_broken_connection(ngx_http_request_t *r,
     ngx_event_t *ev);
+static void ngx_http_upstream_connect(ngx_http_request_t *r,
+    ngx_http_upstream_t *u);
 static ngx_int_t ngx_http_upstream_reinit(ngx_http_request_t *r,
     ngx_http_upstream_t *u);
 static void ngx_http_upstream_send_request(ngx_http_request_t *r,
@@ -432,13 +434,12 @@
 
 found:
 
-    switch(uscf->peer.init(r, uscf)) {
-      case NGX_OK:
-        ngx_http_upstream_connect(r, u);
-      case NGX_BUSY:
+    if (uscf->peer.init(r, uscf) != NGX_OK) {
+        ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
         return;
     }
-    ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
+
+    ngx_http_upstream_connect(r, u);
 }
 
 
@@ -637,7 +638,7 @@
 }
 
 
-void
+static void
 ngx_http_upstream_connect(ngx_http_request_t *r, ngx_http_upstream_t *u)
 {
     ngx_int_t          rc;
diff -Naur src/http/ngx_http_upstream.h ../nginx-0.6.32/src/http/ngx_http_upstream.h
--- src/http/ngx_http_upstream.h	2008-11-09 20:19:11.000000000 +0100
+++ ../nginx-0.6.32/src/http/ngx_http_upstream.h	2008-03-03 21:04:06.000000000 +0100
@@ -270,7 +270,6 @@
     ngx_http_variable_value_t *v, uintptr_t data);
 
 void ngx_http_upstream_init(ngx_http_request_t *r);
-void ngx_http_upstream_connect(ngx_http_request_t *r, ngx_http_upstream_t *u);
 ngx_http_upstream_srv_conf_t *ngx_http_upstream_add(ngx_conf_t *cf,
     ngx_url_t *u, ngx_uint_t flags);
 ngx_int_t ngx_http_upstream_hide_headers_hash(ngx_conf_t *cf,
