diff -Naur src/http/ngx_http_upstream.h src-hash/http/ngx_http_upstream.h
--- src/http/ngx_http_upstream.h	2008-03-03 12:04:06.000000000 -0800
+++ src-hash/http/ngx_http_upstream.h	2008-06-07 13:01:37.000000000 -0700
@@ -90,6 +90,10 @@
 
     ngx_array_t                    *servers;   /* ngx_http_upstream_server_t */
 
+    ngx_array_t                    *values;
+    ngx_array_t                    *lengths;
+    ngx_uint_t                      retries;
+
     ngx_uint_t                      flags;
     ngx_str_t                       host;
     u_char                         *file_name;
